rust/lance-datafusion/src/projection.rs

```diff
@@ -15,85 +15,88 @@ use std::{
 
 use lance_core::{
     datatypes::{OnMissing, Projectable, Projection, Schema},
-    Error, Result,
+    Error, Result, ROW_ADDR, ROW_ID, ROW_OFFSET,
 };
 
 use crate::{
     exec::{execute_plan, LanceExecutionOptions, OneShotExec},
     planner::Planner,
 };
 
+#[derive(Clone, Debug)]
+pub struct OutputColumn {
+    /// The expression that represents the output column
+    pub expr: Expr,
+    /// The name of the output column
+    pub name: String,
+}
+
 #[derive(Clone, Debug)]
 pub struct ProjectionPlan {
-    /// The base thing we are projecting from (e.g. a dataset)
-    base: Arc<dyn Projectable>,
     /// The physical schema that must be loaded from the dataset
     pub physical_projection: Projection,
 
-    /// True if the user wants the row id in the final output
-    ///
-    /// Note: this is related, but slightly different, to physical_projection.with_row_id
-    /// which only tracks if the row id is needed.
-    ///
-    /// desires_row_id implies with_row_id is true
-    /// However, it is possible to have desires_row_id=false and with_row_id=true (e.g. when
-    /// the row id is needed to perform a late materialization take)
-    pub desires_row_id: bool,
-    /// True if the user wants the row address in the final output
-    ///
-    /// Note: this is related, but slightly different, to physical_projection.with_row_addr
-    /// which only tracks if the row address is needed.
-    ///
-    /// desires_row_addr implies with_row_addr is true
-    /// However, it is possible to have deisres_row_addr=false and with_row_addr=true (e.g. during
-    /// a count query)
-    pub desires_row_addr: bool,
+    /// Needs the row address converted into a row offset
+    pub needs_row_offset: bool,
 
-    /// If present, expressions that represent the output columns.  These expressions
-    /// run on the output of the physical projection.
-    ///
-    /// If not present, the output is the physical projection.
-    ///
-    /// Note: this doesn't include _distance, and _rowid
-    pub requested_output_expr: Option<Vec<(Expr, String)>>,
+    /// The desired output columns
+    pub requested_output_expr: Vec<OutputColumn>,
 }
 
 impl ProjectionPlan {
-    /// Create a new projection plan which projects all columns and does not include any expressions
-    pub fn new(base: Arc<dyn Projectable>) -> Self {
-        let physical_projection = Projection::full(base.clone());
-        Self {
-            base,
-            physical_projection,
-            requested_output_expr: None,
-            desires_row_addr: false,
-            desires_row_id: false,
-        }
+    fn add_system_columns(schema: &ArrowSchema) -> ArrowSchema {
```

Comment:
I find `Self::add_system_columns(&full_schema)` is a bit confusing. Is it a good idea to move it out of `ProjectionPlan` as a standalone funcation with new like `add_system_columns_in_schema`?